File: services/compliance-service/src/checker.py
from __future__ import annotations

import re
import sys
from pathlib import Path
from typing import Any

KNOWLEDGE_SRC = Path(__file__).resolve().parents[2] / "knowledge-service" / "src"
sys.path.append(str(KNOWLEDGE_SRC))

from simple_yaml import load_file  # noqa: E402

EXTERNAL_ROLES = {"visitor", "student", "parent"}
PHONE_PATTERN = re.compile(r"1[3-9]\d{9}")


def evaluate_answer(answer: str, scope: dict[str, Any], project_root: Path) -> dict[str, Any]:
    rules = load_file(project_root / "configs" / "compliance_rules.yaml")
    role = scope.get("role", "visitor")
    external_role = role in EXTERNAL_ROLES
    violations: list[str] = []

    blocked_phrases = list(rules.get("blocked_phrases", []))
    if not external_role:
        blocked_phrases = [phrase for phrase in blocked_phrases if phrase not in {"优惠底价", "内部名额"}]

    violations.extend(phrase for phrase in blocked_phrases if phrase in answer)

    if external_role and any(term in answer for term in ["内部参考", "内部规则", "内部话术", "内部优惠"]):
        violations.append("internal_reference_leak")

    if PHONE_PATTERN.search(answer):
        violations.append("privacy_phone_number")

    if "100%" in answer or "一定会" in answer:
        violations.append("absolute_claim")

    guidance = _build_guidance(violations, rules)
    return {
        "passed": not violations,
        "violations": violations,
        "rewrite_guidance": guidance,
    }


def rewrite_answer(answer: str, violations: list[str]) -> str:
    if "privacy_phone_number" in violations:
        return (
            "这个问题涉及个人隐私信息，当前不能直接展示联系方式或学生识别信息。"
            "建议由顾问在授权场景下继续跟进。"
        )

    if "internal_reference_leak" in violations or "优惠底价" in violations:
        return (
            "这个问题需要按学校正式口径处理，不能透露未公开的优惠、内部规则或内部参考内容。"
            "可以先根据学生基础、目标、薄弱科目和意向校区做评估，再由顾问提供合规说明。"
        )

    return (
        "这个问题需要按学校正式口径处理，不能承诺固定提分、保证录取，"
        "也不能透露未公开的优惠或内部规则。可以先根据学生基础、目标、薄弱科目和意向校区做评估，"
        "再由顾问提供合规说明。"
    )


def _build_guidance(violations: list[str], rules: dict[str, Any]) -> list[str]:
    guidance_map = rules.get("rewrite_guidance", {})
    guidance: list[str] = []
    if any(v in violations for v in ["保证录取", "保证提分", "absolute_claim"]):
        guidance.append(str(guidance_map.get("promise_risk", "")))
    if any(v in violations for v in ["优惠底价", "internal_reference_leak"]):
        guidance.append(str(guidance_map.get("pricing_risk", "")))
    if "privacy_phone_number" in violations:
        guidance.append(str(guidance_map.get("privacy_risk", "")))
    return [item for item in guidance if item]

File: services/compliance-service/ts-src/checker.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { parse } from "yaml";

const EXTERNAL_ROLES = new Set(["visitor", "student", "parent"]);
const PHONE_PATTERN = /1[3-9]\d{9}/;

export interface ComplianceResult {
  passed: boolean;
  violations: string[];
  rewrite_guidance: string[];
}

export function evaluateAnswer(
  answer: string,
  scope: Record<string, unknown>,
  projectRoot: string,
): ComplianceResult {
  const rulesPath = resolve(projectRoot, "configs", "compliance_rules.yaml");
  const raw = readFileSync(rulesPath, "utf-8");
  const rules = parse(raw) as Record<string, unknown>;
  const role = (scope.role as string | undefined) ?? "visitor";
  const externalRole = EXTERNAL_ROLES.has(role);
  const violations: string[] = [];

  const rawBlocked = (rules.blocked_phrases ?? []) as string[];
  const blockedPhrases: string[] = [...rawBlocked];
  if (!externalRole) {
    const filtered = blockedPhrases.filter(
      (p) => p !== "优惠底价" && p !== "内部名额",
    );
    violations.push(...filtered.filter((p) => answer.includes(p)));
  } else {
    violations.push(...blockedPhrases.filter((p) => answer.includes(p)));
  }

  if (externalRole) {
    const leakTerms = ["内部参考", "内部规则", "内部话术", "内部优惠"];
    if (leakTerms.some((t) => answer.includes(t))) {
      violations.push("internal_reference_leak");
    }
  }

  if (PHONE_PATTERN.test(answer)) {
    violations.push("privacy_phone_number");
  }

  if (answer.includes("100%") || answer.includes("一定会")) {
    violations.push("absolute_claim");
  }

  const guidance = buildGuidance(violations, rules);
  return {
    passed: violations.length === 0,
    violations,
    rewrite_guidance: guidance,
  };
}

function buildGuidance(
  violations: string[],
  rules: Record<string, unknown>,
): string[] {
  const guidanceMap = (rules.rewrite_guidance ?? {}) as Record<string, unknown>;
  const guidance: string[] = [];

  const promiseViolations = new Set(["保证录取", "保证提分", "absolute_claim"]);
  if (violations.some((v) => promiseViolations.has(v))) {
    guidance.push(String(guidanceMap.promise_risk ?? ""));
  }

  const pricingViolations = new Set(["优惠底价", "internal_reference_leak"]);
  if (violations.some((v) => pricingViolations.has(v))) {
    guidance.push(String(guidanceMap.pricing_risk ?? ""));
  }

  if (violations.includes("privacy_phone_number")) {
    guidance.push(String(guidanceMap.privacy_risk ?? ""));
  }

  return guidance.filter((item) => item !== "");
}

export function rewriteAnswer(answer: string, violations: string[]): string {
  if (violations.includes("privacy_phone_number")) {
    return (
      "这个问题涉及个人隐私信息，当前不能直接展示联系方式或学生识别信息。" +
      "建议由顾问在授权场景下继续跟进。"
    );
  }

  if (violations.includes("internal_reference_leak") || violations.includes("优惠底价")) {
    return (
      "这个问题需要按学校正式口径处理，不能透露未公开的优惠、内部规则或内部参考内容。" +
      "可以先根据学生基础、目标、薄弱科目和意向校区做评估，再由顾问提供合规说明。"
    );
  }

  return (
    "这个问题需要按学校正式口径处理，不能承诺固定提分、保证录取，" +
    "也不能透露未公开的优惠或内部规则。可以先根据学生基础、目标、薄弱科目和意向校区做评估，" +
    "再由顾问提供合规说明。"
  );
}

export function evaluateAndRewrite(
  answer: string,
  scope: Record<string, unknown>,
  projectRoot: string,
): ComplianceResult & { rewritten_answer: string } {
  const result = evaluateAnswer(answer, scope, projectRoot);
  return {
    ...result,
    rewritten_answer: result.passed ? answer : rewriteAnswer(answer, result.violations),
  };
}
